zanata-war/src/main/java/org/zanata/rest/editor/service/SuggestionsService.java

```diff
@@ -0,0 +1,142 @@
+
+import static javax.ws.rs.core.Response.Status.BAD_REQUEST;
+import static org.zanata.webtrans.shared.rpc.HasSearchType.*;
+
+/**
+ * @see org.zanata.rest.editor.service.resource.SuggestionsResource
+ */
+@Name("editor.suggestionsService")
+@Path(SuggestionsResource.SERVICE_PATH)
+@Transactional
+public class SuggestionsService implements SuggestionsResource {
+
+    public static final String SEARCH_TYPES = Joiner.on(", ").join(SearchType.values());
+
+    @In("translationMemoryServiceImpl")
+    private TranslationMemoryService transMemoryService;
+
+    @In("localeServiceImpl")
+    private LocaleService localeService;
+
+    @Override
+    public Response query(List<String> query, String sourceLocaleString, String transLocaleString, String searchTypeString) {
+
+        Either<SearchType, Response> searchTypeOrErrorResponse = getSearchType(searchTypeString);
+        if (searchTypeOrErrorResponse.isRight()) {
+            return searchTypeOrErrorResponse.right();
+        }
+        SearchType searchType = searchTypeOrErrorResponse.left();
+
+        Either<LocaleId, Response> sourceLocaleOrError = getSourceLocale(sourceLocaleString);
+        if (sourceLocaleOrError.isRight()) {
+            return sourceLocaleOrError.right();
+        }
+        LocaleId sourceLocale = sourceLocaleOrError.left();
+
+        Either<LocaleId, Response> transLocaleOrError = getTransLocale(transLocaleString);
+        if (transLocaleOrError.isRight()) {
+            return transLocaleOrError.right();
+        }
+        LocaleId transLocale = transLocaleOrError.left();
+
+        List<Suggestion> suggestions = transMemoryService.searchTransMemoryWithDetails(transLocale, sourceLocale,
+                new TransMemoryQuery(query, searchType));
+
+        // Wrap in generic entity to prevent type erasure, so that an
+        // appropriate MessageBodyReader can be used.
+        // see docs for GenericEntity
+        GenericEntity<List<Suggestion>> entity = new GenericEntity<List<Suggestion>>(suggestions) {};
+
+        return Response.ok(entity).build();
+    }
+
+    private Either<LocaleId, Response> getSourceLocale(String localeString) {
+        return getLocaleOrError(localeString, "Unrecognized source locale: \"%s\"");
+    }
+
+    private Either<LocaleId, Response> getTransLocale(String localeString) {
+        return getLocaleOrError(localeString, "Unrecognized translation locale: \"%s\"");
+    }
+
+    /**
+     * Try to get a valid locale for a given string. If the string does not match a supported
+     * locale, generate an error response with the given error message.
+     *
+     * @param localeString used to look up the locale
+     * @param errorFormat error to display. Include "%s" to substitute the locale string.
+     * @return a LocaleId if the given string matches one, otherwise an error response.
+     */
+    private Either<LocaleId, Response> getLocaleOrError(final String localeString, final String errorFormat) {
```

Comment:
> By making a format string one of the parameters, String.format already influenced the semantics of the method, so why not let it influence the signature?

I changed it so that is passes an already-formatted string instead, once I considered how error-prone it might be to have the format string distant from the call to `String.format`